projection/proofs/to_cake/payload_to_cakemlProofScript.sml

```diff
@@ -2822,29 +2838,84 @@ Theorem network_forward_correctness:
-   >- (asm_exists_tac \\ fs []
-       \\ asm_exists_tac \\ fs []
-       \\ irule ffi_irrel \\ fs []
-       \\ conj_tac
-       >- metis_tac [ffi_wf_def,trans_ffi_eq_same]
-       \\ MAP_EVERY qexists_tac [‘p’,‘s’]
-       \\ rw [cpEval_valid_def,ffi_state_cor_def]
-       >- rw [ffi_wf_def]
-       \\ irule internal_trans_pres_wf
-       \\ MAP_EVERY qexists_tac [‘n’,‘conf’,‘s.queues’]
-       \\ rw [ffi_wf_def] \\ irule RTC_SINGLE
-       \\ rw [comms_ffi_consTheory.internal_trans_def]
-       \\ last_x_assum (assume_tac o ONCE_REWRITE_RULE [trans_cases]) \\ fs []
-       \\ IMP_RES_TAC trans_not_same \\ rw [] \\ fs [])
-   \\ drule_then (qspecl_then [‘vs1’,‘env1’,‘st1’] mp_tac) endpoint_forward_correctness
-   \\ CONV_TAC (PAT_CONV “λx. (∀vs2. x) ⇒ _” SWAP_FORALL_CONV)
-   \\ CONV_TAC (PAT_CONV “λx. x ⇒ _” SWAP_FORALL_CONV)
-   \\ disch_then (qspec_then ‘st2’ mp_tac)
-   \\ CONV_TAC (PAT_CONV “λx. x ⇒ _” SWAP_FORALL_CONV)
-   \\ cheat
+  rw []
+  \\ drule_then assume_tac NPar_trans_l_cases
+  \\ fs [] \\ rveq
+  (* p is not involved at all *)
+  >- (CONV_TAC SWAP_EXISTS_CONV
+      \\ asm_exists_tac \\ fs []
+      \\ asm_exists_tac \\ fs []
+      \\ irule ffi_irrel \\ fs []
+      \\ conj_tac
+      >- metis_tac [ffi_wf_def,trans_ffi_eq_same]
+      \\ MAP_EVERY qexists_tac [‘p’,‘s’]
+      \\ rw [cpEval_valid_def,ffi_state_cor_def]
+      >- rw [ffi_wf_def]
+      \\ irule internal_trans_pres_wf
+      \\ MAP_EVERY qexists_tac [‘n’,‘conf’,‘s.queues’]
+      \\ rw [ffi_wf_def] \\ irule RTC_SINGLE
+      \\ rw [comms_ffi_consTheory.internal_trans_def]
+      \\ last_x_assum (assume_tac o ONCE_REWRITE_RULE [trans_cases]) \\ fs []
+      \\ IMP_RES_TAC trans_not_same \\ rw [] \\ fs [])
+  (* LTau (only p does something) *)
+  >- (drule_then (qspecl_then [‘vs1’,‘env1’,‘st1’,‘st2’] mp_tac) endpoint_forward_correctness
+      \\ impl_tac
+      >- (rw [cpEval_valid_def,ffi_wf_def,
+              ffi_state_cor_def,
+              cpFFI_valid_def,some_def]
+          \\ fs []
+          \\ qpat_assum `trans _ (NEndpoint _ _ _) _ _`
+                          (mp_tac o PURE_ONCE_REWRITE_RULE [trans_cases])
+          \\ fs [] \\ rw []
+          \\ fs [state_component_equality] \\ rveq \\ rfs [state_component_equality]
+          >- (SELECT_ELIM_TAC \\ rw []
+              >- (asm_exists_tac \\ fs [])
+              \\ Cases_on ‘x'’ \\ fs []
+              \\ Cases_on ‘tl’ \\ rw [] >- cheat
+              \\ fs [normalise_queues_FUPDATE_NONEMPTY]
+              \\ cheat)
+          >- cheat
+          >- (Cases_on ‘x’ \\ fs []
+              \\ qpat_x_assum ‘s.queues = _’ (assume_tac o Q.AP_TERM ‘λe. qlk e q’)
+              \\ fs [])
+          >- (Cases_on ‘x’ \\ fs []
+              \\ qpat_x_assum ‘s.queues = _’ (assume_tac o Q.AP_TERM ‘λe. qlk e q’)
+              \\ fs [])
+          >- (Cases_on ‘x’ \\ fs []
+              \\ qpat_x_assum ‘s.queues = _’ (assume_tac o Q.AP_TERM ‘λe. qlk e q’)
+              \\ fs [])
+          >- cheat
```

Comment:
This one is just a weird case where a dequeue is attempted but the queue is not in the right shape, I think it can be discarded  by contradiction, but I have try it yet

Reply:
I think deriving a contradiction here might require having `normalised(s.queues)` in the assumptions. (not a problem since this is an invariant that `trans` maintains, and `strans` should maintain)